collector/system_collector.go

```diff
@@ -0,0 +1,152 @@
+		"Status of NSX-T system cluster nodes UP/DOWN.",
+		[]string{"name", "type"},
+		nil,
+	)
+	applianceManagementServiceStatus := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "appliance_service", "status"),
+		"Status of NSX-T appliance management service UP/DOWN.",
+		nil,
+		nil,
+	)
+	return &systemCollector{
+		systemClient: nsxtClient,
+		logger:       logger,
+
+		clusterStatus:                    clusterStatus,
+		clusterNodeTotal:                 clusterNodeTotal,
+		clusterNodeUp:                    clusterNodeUp,
+		clusterNodeStatus:                clusterNodeStatus,
+		applianceManagementServiceStatus: applianceManagementServiceStatus,
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (sc *systemCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- sc.clusterStatus
+	ch <- sc.clusterNodeTotal
+	ch <- sc.clusterNodeUp
+	ch <- sc.clusterNodeStatus
+	ch <- sc.applianceManagementServiceStatus
+}
+
+// Collect implements the prometheus.Collector interface.
+func (sc *systemCollector) Collect(ch chan<- prometheus.Metric) {
+	sc.collectClusterStatusMetric(ch)
+	sc.collectClusterNodeMetric(ch)
+	sc.collectClusterServiceMetric(ch)
+}
+
+func (sc *systemCollector) collectClusterStatusMetric(ch chan<- prometheus.Metric) {
+	clusterStatus, err := sc.systemClient.ReadClusterStatus()
+	if err != nil {
+		level.Error(sc.logger).Log("msg", "Unable to collect cluster status")
+		return
+	}
+
+	if strings.ToUpper(clusterStatus.ControlClusterStatus.Status) == "STABLE" &&
+		strings.ToUpper(clusterStatus.MgmtClusterStatus.Status) == "STABLE" {
+		ch <- prometheus.MustNewConstMetric(sc.clusterStatus, prometheus.GaugeValue, 1.0)
+	} else {
+		ch <- prometheus.MustNewConstMetric(sc.clusterStatus, prometheus.GaugeValue, 0.0)
+	}
+
+	clusterNodeUp := len(clusterStatus.MgmtClusterStatus.OnlineNodes)
+	clusterNodeDown := len(clusterStatus.MgmtClusterStatus.OfflineNodes)
+	clusterNodeTotal := clusterNodeUp + clusterNodeDown
+	ch <- prometheus.MustNewConstMetric(sc.clusterNodeTotal, prometheus.GaugeValue, float64(clusterNodeTotal))
+	ch <- prometheus.MustNewConstMetric(sc.clusterNodeUp, prometheus.GaugeValue, float64(clusterNodeUp))
+}
+
+func (sc *systemCollector) collectClusterNodeMetric(ch chan<- prometheus.Metric) {
+	clusterNodesStatus, err := sc.systemClient.ReadClusterNodesAggregateStatus()
+	if err != nil {
+		level.Error(sc.logger).Log("msg", "Unable to collect cluster nodes status")
+		return
+	}
+
+	for _, c := range clusterNodesStatus.ControllerCluster {
+		// No node name property provided, use ip address instead
+		name := c.RoleConfig.ControlPlaneListenAddr.IpAddress
+
+		if strings.ToUpper(c.NodeStatus.ControlClusterStatus.ControlClusterStatus) == "CONNECTED" &&
+			strings.ToUpper(c.NodeStatus.ControlClusterStatus.MgmtConnectionStatus.ConnectivityStatus) == "CONNECTED" {
+			ch <- prometheus.MustNewConstMetric(sc.clusterNodeStatus, prometheus.GaugeValue, 1.0, name, "controller")
```

Comment:
For constant "controller" and "management" should we create two different desc and use constant labels instead? It's in the fourth parameter of NewDesc

Reply:
good idea, let me do that